### ISSUE TITLE
Does lozad try and load lazyload images during idle time?

### ISSUE THREAD

**Issue by sketchbuch** [ApoorvSaxena/lozad.js#110]

**Comment by ApoorvSaxena** [comment]
Nope, but this looks like a good addition. How do you plan to detect idle time in a performant manner?

**Comment by sketchbuch** [comment]
I was hoping you would come up with a performant way ;)

Well obviously to start with use window.requestIdleCallback() for supporting browsers... sadly this doesn't include Edge or Safari.
What about dynamically adding <link rel="prefetch"...> tags to the dom? Won't they then automatically be downloaded during idle time... and then when lozad tries to lazy load a preloaded image it will won't need downloading.

**Comment by Jab2870** [comment]
I'd say that this is an enhancement so (as long as it doesn't break anything) browser support shouldn't be a huge concern.

Browsers that support requestIdleCallback will get the added performance, those that don't won't.

FYI, it is likely that future releases of edge will support it:
https://developer.microsoft.com/en-us/microsoft-edge/platform/status/requestidlecallback/
and Safari is considering it:
https://webkit.org/status/#feature-requestidlecallback

**Comment by sketchbuch** [comment]
Maybe I shall get round this week or next week to making a pull request

**Comment by ApoorvSaxena** [comment]
we should better make this configurable, as we may not want to consume user's bandwidth unnecessarily in situations like mobile browser and less capable network connection

**Comment by allanwhite** [comment]
@sketchbuch How can you dynamically add a prefetch call to the dom, in such a way that it behaves like prefetch should? It would seem to me, that by the time you've inserted that (after initial load, initial dom parse, then JS parse) the utility of prefetch has passed. For items way down the page, perhaps. 

I'd think it would make more sense if image URLs are available to a page/template, baking those up in the head as rel preloads would be most effective (rather than waiting for JS to get around to injecting them). Good reminder to leverage preloads though.

**Comment by ApoorvSaxena** [comment]
our discussed approach got exactly implemented recently by a [GoogleChromeLabs project](https://github.com/GoogleChromeLabs/quicklink#how-it-works), where they:
- **Waits until the browser is idle** (using [requestIdleCallback](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback))
- **Checks if the user isn't on a slow connection** (using navigator.connection.effectiveType) or has data-saver enabled (using navigator.connection.saveData)
- **Prefetches URLs to the links** (using [\<link rel=prefetch\>](https://www.w3.org/TR/resource-hints/#prefetch) or XHR). Provides some control over the request priority (can switch to fetch() if supported)

seems like a validated move now for implementation ;)
